src/kibana/components/vislib/lib/legend.js

```diff
@@ -17,16 +20,21 @@ define(function (require) {
      * @param color {Function} Color function
      * @param _attr {Object|*} Reference to Vis options
      */
-    function Legend(vis, el, labels, color, _attr) {
+    function Legend(vis) {
       if (!(this instanceof Legend)) {
-        return new Legend(vis, el, labels, color, _attr);
+        return new Legend(vis);
       }
 
+      var data = vis.data.columns || vis.data.rows || [vis.data];
+      var type = vis._attr.type;
+      var labels = this.labels = this._getLabels(data, type);
+      var labelsArray = labels.map(function (obj) { return obj.label; });
+
+      this.events = new Dispatch();
       this.vis = vis;
-      this.el = el;
-      this.labels = labels;
-      this.color = color;
-      this._attr = _.defaults(_attr || {}, {
+      this.el = vis.el;
+      this.color = color(labelsArray);
+      this._attr = _.defaults(vis._attr || {}, {
```

Comment:
This may mutate the vis._attr object. You can use `_.defaults({}, vis._attr || {}, ...)` to avoid that side effect